src/witan/send/adroddiad/clerk/charting_v2.clj

```diff
@@ -0,0 +1,147 @@
+
+
+
+(defn color-map [plot-data color-field color-lookup]
+  (let [group-keys (into (sorted-set) (get plot-data color-field))
+        filtered-colors (tc/select-rows color-lookup #(group-keys (get % :domain-value)))]
+    {:field color-field
+     :scale {:range (into [] (:color filtered-colors))
+             :domain (into [] (:domain-value filtered-colors))}}))
+
+(defn shape-map [plot-data shape-field shape-lookup]
+  (let [group-keys (into (sorted-set) (get plot-data shape-field))
+        filtered-shapes (tc/select-rows shape-lookup #(group-keys (get % :domain-value)))]
+    {:field shape-field
+     :scale {:range (into [] (:shape filtered-shapes))
+             :domain (into [] (:domain-value filtered-shapes))}}))
+
+(defn line-plot
+  [{:keys [data
+           chart-title
+           chart-height chart-width
+           clerk-width
+           x x-title x-format
+           y y-title y-format y-zero
+           group group-title
+           colors-and-shapes]
+    :or {chart-height 500
+         chart-width 500
+         clerk-width :full
+         y-zero true}}]
+  (let [tooltip [{:field group, :title group-title},
+                 {:field x, :type "temporal", :format x-format, :title x-title},
+                 {:field y, :title y-title ;; :format y-format
+                  }]]
+    (clerk/vl
+     {::clerk/width clerk-width}
+     {:height chart-height
+      :width chart-width
+      ;; :autosize {:type "fit" :contains "padding"}
+      :title {:text chart-title
+              :fontSize 24}
+      :config {:legend {:titleFontSize 16
+                        :labelFontSize 14}
+               :axisX {:tickcount 7
+                       :tickExtra true
+                       :labelalign "center"
+                       :titleFontSize 16
+                       :labelFontSize 12}
+               :axisY {:titleFontSize 16
+                       :labelFontSize 12}}
+      :data {:values (-> data
+                         (tc/rows :as-maps))}
+      :layer [{:mark {:type "line", :point {:filled false,
+                                            :fill "white",
+                                            :size 50
+                                            :strokewidth 0.5}},
+               :encoding {:y {:field y, :title y-title :type "quantitative" :scale {:zero y-zero}},
+                          :x {:field x, :title x-title :format x-format :type "temporal"},
+                          ;; color and shape scale and range must be specified or you get extra things in the legend
+                          :color (assoc (color-map data group colors-and-shapes) :title group-title)
+                          :shape (shape-map data group colors-and-shapes)
+                          :tooltip tooltip}}]})))
+
+(defn line-and-ribbon-plot
+  [{:keys [data
+           chart-title
+           chart-height chart-width
+           clerk-width
+           x x-title x-format
```

Comment:
What should `x-format` be?

Reply:
what ever is appropriate to format the numbers on the axis (say 10.2M for GBP)